### PR TITLE
New App: First pass at the Islamic prayer times app

### PR DESCRIPTION
**Pull request by austinrfnd**
islamic_prayer

### CODE REVIEW

**Comment by zeronullempty** [comment]
Awesome work! I would recommend a few things.

Sunrise, Sunset, Midnight, and Ishraq don't need to be there. 5 daily prayers (fajr, zuhr, asr, maghrib, isha) is really all that's needed. Sunset is actually exactly the same as Maghrib! 

Apart from that, awesome job!!!

**Comment by austinrfnd** [comment]
@zeronullempty Love the feedback.  Thank you! You think it should be configurable to show/not show the others?  If so, any suggestions on what to name the configurable setting? Maybe something like `expanded_times`?

**Comment by zeronullempty** [comment]
I think that a sunrise toggle might be useful for a lot of people (it indicates when the time for Fajr ends)

**Comment by zeronullempty** [comment]
Also, to add. I just had a chance to push your app to my Tidbyt. Scrolling is **very** fast, almost too fast to read.

**Comment by austinrfnd** [comment]
> had

Cool, good to know. I'll let you know when I push an update.  I would love any additional feedback.

**Comment by austinrfnd** [comment]
@zeronullempty can you give it another round of tries?  Let me know any additional feedback.  Thanks again for all your help!

**Comment by zeronullempty** [comment]
Sure. A few things I noticed.

1) Scrolling is still a bit fast. Biggest weird thing is that the text starts from the bottom and then scrolls all the way up. Seems a *little* unnecessary. Maybe you could start off with showing all the times that would fit, then slowly scroll down to show the last one.  Not sure what the best way to do this would be, honestly, someone on the forums suggested to maybe ONLY show future times? The API has a flag you can set for iso8601 times, I don't know if that plays well with starlark's time library.

I just feel like for something that is cycling so often, especially in many people'  field of view (on a desk, etc) you'd want to minimize movement. Could just be my personal preference though. 

2) On the topic of APIs... You seem to be using the calendar version of the API, fetching all the timings for the current month, then caching them for a month.  What if there's a new month? The route I was thinking of implementing was fetching new times daily, using the "Timings" version of the API (\#11 on the page), then caching them until the end of the day, or for a few hours at least.


Also, I don't think you *need* to include options for the prayer calculation methods. The option is nice, but most people don't really know exactly what method is being used where they are. The API actually is able to pick the correct one based on the location provided, if no argument is provided. So maybe have a "default" calculation in the schema that accounts for this? This way, people can get a calculation, but if there's a specific one they want to use, they have the option.

Sorry for the wall of text, this is my first time with open source, not really sure what amount of text is expected.

**Comment by zeronullempty** [comment]
Also, 24 hour vs 12 hour toggle?

**Comment by austinrfnd** [comment]
@zeronullempty zero (get it?) apologizes needed! Seriously great feedback.  I got off my lazy horse and actually pushed to my tidbyt saw the display issues.  I gave it another go.  So pull down these changes and give it another try.  As for only display prayer times in the future.  I call it "future time mode", I added it as a todo.  I'll get around to it I just need to brush on Starlark's time library.

As far the API is concerned.  I read the API docs.  As for now the default is to throw nothing to the API call.  So by default it behaves by calculating the preferred method.  And then allows the override.  Maybe I'll update the description of that setting to better show the intention and how I use that API parameter.  But we are aligned in this point and by default the app does behave aligned with your feedback.

As far as the month API call.  I think I have that covered where the cache key is the "month/year".  So whenever a new month rolls around that cache entry isn't there.  So it should be fine... but yeah I've been confident in my solutions before for it to backfire LOL.  

Thanks again for the feedback.  It's been awesome working with you!  Give it a shot and let me know any feedback.

**Comment by zeronullempty** [comment]
Awesome work! Looks great. 

The colors on the first 2 timings are a bit hard to read, at least for me. Wouldn't mind an option to switch to all-white text. Ditto for the top header.

And a small suggestion in the schema, "Show Sunrise Prayer" should be just "Show Sunrise Time" since there isn't actually a specific sunrise prayer. It basically signifies the cutoff for the Fajr prayer.

All in all, awesome job man, its basically perfectly usable at the moment, hopefully people can make use of this soon.

**Review by betterengineering — COMMENTED** [review]
This looks great! Thanks so much @austinrfnd and @zeronullempty! It's so awesome to see collaboration on a PR like this 🙌 .

File: apps/islamicprayer/islamic_prayer.star [hunk 1]
```diff
@@ -0,0 +1,306 @@
...
+    return filtered_prayer_times
+
+def fetch_prayer_times(latitude, longitude, day, month, year, prayer_calc_option):
+    cache_key = "prayer_{}/{}".format(month, year)
+
+    cached_data = cache.get(cache_key)
+
+    if cached_data == None:
+        # API docs: https://aladhan.com/prayer-times-api#GetCalendar
+        api_url = "http://api.aladhan.com/v1/calendar?latitude={}&longitude={}&month={}&year={}&method={}".format(latitude, longitude, month, year, prayer_calc_option)
```
[the remark this answers is not in the document]
Comment (austinrfnd) [reply apps/islamicprayer/islamic_prayer.star:183]: @betterengineering yeah no problem.  I'll take care of this soon.

[the remark this answers is not in the document]
Comment (austinrfnd) [reply apps/islamicprayer/islamic_prayer.star:189]: @betterengineering done.  Thanks for the heads up.

**Review by betterengineering — APPROVED** [review]

**Pull request merged**